File: trunk/server/examples/plugins/console/js/console.js
/**
  * Himle Server -- http://himle.org/
  *
  * Copyright (C) 2008 Juha-Jarmo Heinonen
  *
  * This file is part of Himle Server.
  *
  * Himle Server is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
  *
  * Himle server is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  *
  **/

HTMLViewer = HControl.extend({
  flexRight: true,
  flexBottom: true,
  draw: function(){
    this.base();
    var i=0, _styles = [
      ['font-family','monaco, monospace, courier new'],
      ['font-size','9px'],
      ['color','#333'],
      ['overflow','auto']
    ];
    for(;i<_styles.length;i++){
      this.setStyle(_styles[i][0],_styles[i][1]);
    }
  },
  setValue: function(_value){
    this.base(_value);
    this.setHTML('<pre>'+_value+'</pre>');
  }
});


HConsole = HApplication.extend({
  constructor: function(_values){
    this.values = _values;
    this.base();
    this.view = new HDynControl(
      new HRect(200,20,200+512,100+384),
      this, {
        label: '',
        resizeN: 4, resizeS: 4, resizeE: 4, resizeW: 4,
        resizeNE: [4,4], resizeNW: [4,4], resizeSE: [4,4], resizeSW: [4,4]
      }
    );
    this.view.setStyle('background-color','#ddd');
    this.tabs = new (HTab.extend(
      {flexBottom:true,flexRight:true,flexBottomOffset:4,flexRightOffset:4}
    ))(
      new HRect(4,24,512-4,384-4),
      this.view
    );
    var _sessionTab  = this.tabs.addTab('Sessions',true);
    var _imgServeTab = this.tabs.addTab('ImageServe');
    
    var _htmlViewer  = new HTMLViewer(new HRect(0,0,1,1),_sessionTab);
    HVM.values[this.values.session_to_js].bind(_htmlViewer);
    
    var _reloadButtonRect = new HRect(4,4,100,20);
    var _reloadButton = new HClickButton(_reloadButtonRect,this.view,{label:'Refresh'});
    HVM.values[this.values.reload_button].bind(_reloadButton);
    
    var _clearButtonRect = new HRect(_reloadButtonRect);
    _clearButtonRect.offsetBy(100,0);
    var _clearButton = new HClickButton(_clearButtonRect,this.view,{label:'Clear'})
    HVM.values[this.values.clear_button].bind(_clearButton)
    
    
  }
});

File: trunk/server/examples/plugins/console/console.rb
# -* coding: UTF-8 -*-
###
  # Himle Server -- http://himle.org/
  #
  # Copyright (C) 2008 Juha-Jarmo Heinonen
  #
  # This file is part of Himle Server.
  #
  # Himle Server is free software: you can redistribute it and/or modify
  # it under the terms of the GNU General Public License as published by
  # the Free Software Foundation, either version 3 of the License, or
  # (at your option) any later version.
  #
  # Himle server is distributed in the hope that it will be useful,
  # but WITHOUT ANY WARRANTY; without even the implied warranty of
  # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  # GNU General Public License for more details.
  #
  # You should have received a copy of the GNU General Public License
  # along with this program.  If not, see <http://www.gnu.org/licenses/>.
  #
  ###

require 'pp'

class HConsole < Plugin
  def hash_to_js( node_hash, node='', indent=0 )
    node += "{ "
    keysize = node_hash.keys.size
    indent += 1
    node_hash.each_key do |key|
      value = node_hash[key]
      unless keysize == node_hash.keys.size
        node += "#{'  '*indent}"
      end
      node += "#{key.inspect} =&gt; "
      keysize -= 1
      if key == :by_id
        value = value.keys
      end
      if value.class == HValue
        node += %{#<HValue:#{value.val_id} @data="#{value.data.inspect[0..100]}#{'...' if value.data.inspect.size > 100}", @sync=#{value.sync.inspect}, @valid=#{value.valid.inspect}, @jstype=#{value.jstype.inspect}, @members=#{value.members.inspect}>}.gsub('<','&lt;').gsub('>','&gt;')
      elsif value.class == Hash
        node += hash_to_js( value, node, indent+1 )
      else
        node += "#{value.inspect[0..100].gsub('<','&lt;').gsub('>','&gt;')}"
      end
      if keysize == 0
        node += "\n"
      else
        node += ",\n"
      end
    end
    indent -= 1
    node += "#{'  '*indent}}\n"
    return node
  end
  def init_ses(msg)
    msg.session[:console] = {
      :session_to_js => HValue.new(msg,''),
      :reload_button => HValue.new(msg,0),
      :clear_button  => HValue.new(msg,0)
    }
    msg.session[:console][:reload_button].bind('console','reload_session_to_js')
    msg.session[:console][:clear_button].bind('console','clear_session_to_js')
  end
  def restore_ses(msg)
    init_ses(msg) unless msg.session.has_key?(:console)
  end
  def reload_session_to_js(msg,val)
    pp msg.session
    msg.session[:console][:session_to_js].set(msg,hash_to_js(msg.session))
    return true
  end
  def clear_session_to_js(msg,val)
    msg.session[:console][:session_to_js].set(msg,'')
    return true
  end
  def idle(msg)
    include_js(msg,['basic','tabs'])
    if msg.session.has_key?(:main)
      if msg.session[:main][:boot] == 2
        msg.reply require_js('console')
        msg.reply "cons = new HConsole({
  session_to_js: '#{msg.session[:console][:session_to_js].val_id}',
  reload_button: '#{msg.session[:console][:reload_button].val_id}',
  clear_button:  '#{msg.session[:console][:clear_button].val_id}'
  
});"
      end
    end
  end
end
console = HConsole.new
console.register('console')